app/controllers/collections_controller.rb

```diff
@@ -24,45 +24,40 @@ def load_collection_from_id
   end
 
   def index
-    if params[:work_id] && (@work = Work.find_by(id: params[:work_id]))
+    if params[:work_id] && (@work = Work.find_by!(id: params[:work_id]))
       @collections = @work.approved_collections.by_title.includes(:parent, :moderators, :children, :collection_preference, owners: [:user]).paginate(page: params[:page])
-    elsif params[:collection_id] && (@collection = Collection.find_by(name: params[:collection_id]))
-      @collections = @collection.children.by_title.includes(:parent, :moderators, :children, :collection_preference, owners: [:user]).paginate(page: params[:page])
-    elsif params[:user_id] && (@user = User.find_by(login: params[:user_id]))
-      @collections = @user.maintained_collections.by_title.includes(:parent, :moderators, :children, :collection_preference, owners: [:user]).paginate(page: params[:page])
+    elsif params[:collection_id] && (@collection = Collection.find_by!(name: params[:collection_id]))
+      @search = CollectionSearchForm.new({ parent_id: @collection.id }.merge(page: params[:page]))
+      @collections = @search.search_results
+    elsif params[:user_id] && (@user = User.find_by!(login: params[:user_id]))
+      @search = CollectionSearchForm.new({ maintainer_id: @user.id }.merge(page: params[:page]))
+      @collections = @search.search_results
       @page_subtitle = ts("%{username} - Collections", username: @user.login)
     else
-      if params[:user_id]
-        flash.now[:error] = ts("We couldn't find a user by that name, sorry.")
-      elsif params[:collection_id]
-        flash.now[:error] = ts("We couldn't find a collection by that name.")
-      elsif params[:work_id]
-        flash.now[:error] = ts("We couldn't find that work.")
-      end
       @sort_and_filter = true
-      params[:collection_filters] ||= {}
-      params[:sort_column] = "collections.created_at" if !valid_sort_column(params[:sort_column], 'collection')
-      params[:sort_direction] = 'DESC' if !valid_sort_direction(params[:sort_direction])
-      sort = params[:sort_column] + " " + params[:sort_direction]
-      @collections = Collection.sorted_and_filtered(sort, params[:collection_filters], params[:page])
+      @search = CollectionSearchForm.new(collection_filter_params.merge(page: params[:page]))
```

Comment:
The nice thing about listing child collections is that it sort of resolves https://otwarchive.atlassian.net/browse/AO3-3748. It also handles challenge type mismatches between parent and child. (The "Gift Exchange Challenge" radio button on the collections page actually excludes some recurring gift exchanges like Yuletide, because only the child collections are marked as a gift exchange, and the child collections aren't listed in production.)

If the search only returns parent collections, I think the solution to AO3-3748 will have to involve adding a field with all child collection names to the index (and maybe all child challenge types, as well). Should that be part of this PR, to save having to reindex again later?

Reply:
Yeah, sorry, I wasn't clear -- I just meant ao3.org/collections in its initial state. I don't particularly mind the change for results.

However, I can see how it would be confusing to have the page do something different when it was filtered than when it wasn't, _and_ it is clearer and more convenient to show the subcollection that matches the search/filter criteria than to show its parent. So maybe it's not a bad change, even if it makes the unfiltered list a bit more cluttered/overwhelming.

If we keep the new behavior, we just need to make sure to change the buttons on the Open Challenges pages that say "Top-Level Collections."